cmd/containerd-shim-runhcs-v1/exec_clone.go

```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"context"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/sirupsen/logrus"
+)
+
+func newClonedExec(ctx context.Context, events publisher, tid, bundle string, io upstreamIO) *clonedExec {
+	log.G(ctx).WithFields(logrus.Fields{
+		"tid":    tid,
+		"eid":    tid, // Init exec ID is always same as Task ID
+		"bundle": bundle,
+	}).Debug("newClonedExec")
+
+	wpse := &wcowPodSandboxExec{
+		events:     events,
+		tid:        tid,
+		bundle:     bundle,
+		state:      shimExecStateCreated,
+		exitStatus: 255, // By design for non-exited process status.
+		exited:     make(chan struct{}),
+	}
+
+	ce := &clonedExec{
+		wpse,
+		io,
+	}
+	return ce
+}
+
+var _ = (shimExec)(&clonedExec{})
+
+// ClonedExec is almost the same as wcowPodSandboxExec in that it doesn't actually track
+// a process (it is just a dummy exec). However, Cloned exec tracks the IO channels passed
+// with the request and it closes them correctly when the exec kill is called.
```

Comment:
Can you expand on this to describe why that change was needed?

Reply:
A ClonedExec is created to track the init process of the cloned container. In reality there is no such init process in the cloned container (as per current design - it might change later). However, the client will still create the stdio pipes for this init process. And when the cloned container is killed we need to correctly close these pipes. Otherwise the client will keep waiting forever for these pipes to be closed. That's why we have this separate ClonedExec type which will close these pipes correctly after the container exits.